src/lib.rs

```diff
@@ -1642,23 +1651,29 @@ pub unsafe trait TryFromBytes {
     /// these cases are handled.
     #[must_use = "has no side effects"]
     #[inline]
-    fn try_mut_from(bytes: &mut [u8]) -> Option<&mut Self>
+    fn try_mut_from(bytes: &mut [u8]) -> Result<&mut Self, TryCastError<&mut [u8]>>
     where
         Self: KnownLayout + Immutable, // TODO(#251): Remove the `Immutable` bound.
     {
-        let candidate = Ptr::from_mut(bytes).try_cast_into_no_leftover::<Self>()?;
-
-        // This call may panic. If that happens, it doesn't cause any soundness
-        // issues, as we have not generated any invalid state which we need to
-        // fix before returning.
-        //
-        // Note that one panic or post-monomorphization error condition is
-        // calling `try_into_valid` (and thus `is_bit_valid`) with a shared
-        // pointer when `Self: !Immutable`. Since `Self: Immutable`, this panic
-        // condition will not happen.
-        let candidate = candidate.try_into_valid();
-
-        candidate.map(Ptr::as_mut)
+        match Ptr::from_mut(bytes).try_cast_into_no_leftover::<Self>() {
+            Ok(candidate) => {
+                // This call may panic. If that happens, it doesn't cause any soundness
+                // issues, as we have not generated any invalid state which we need to
+                // fix before returning.
+                //
+                // Note that one panic or post-monomorphization error condition is
+                // calling `try_into_valid` (and thus `is_bit_valid`) with a shared
+                // pointer when `Self: !Immutable`. Since `Self: Immutable`, this panic
+                // condition will not happen.
+                match candidate.try_into_valid() {
+                    Ok(candidate) => Ok(candidate.as_mut()),
+                    Err(e) => {
+                        Err(e.map_src(|src| src.as_bytes().as_mut()).into()) // TODO: FIXME
```

Comment:
In many places of this PR, returning the remaining bytes required restructuring the function body so the borrow checker can reason better about control flow.

This is the one place where I've yet to figure out how to make the borrow checker happy. The code here doesn't violate the borrow checker, but it fails to compile because `.as_bytes()` is only callable on `Sized` referents. We could make `.as_bytes()` work on `?Sized` referents with `size_of_val_raw`, but that's still yet to be stabilized.